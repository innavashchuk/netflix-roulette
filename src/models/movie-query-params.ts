import { SortingDirectionEnum, SortingFieldsEnum } from './enums/movies-list';

export interface MovieQueryParams {
  toQueryString?: () => string,
  sortBy?: string,
  sortOrder?: string,
  search?: string,
  searchBy?: string,
  filter?: string,
  offset?: string,
  limit?: string
}

export class MovieQueryParamsDict implements MovieQueryParams {

  sortBy?: string;

  sortOrder?: string;

  search?: string;

  searchBy?: string;

  filter?: string;

  offset?: string;

  limit?: string;

  constructor(params: MovieQueryParams = {
    sortBy: '',
    sortOrder: SortingDirectionEnum.desc,
    search: '',
    searchBy: 'title',
    filter: '',
    offset: '',
    limit: '24'
  }) {
    if (!params) {
      return this;
    }
    this.sortBy = params.sortBy || '';
    this.sortOrder = params.sortOrder || SortingDirectionEnum.desc;
    this.search = params.search || '';
    this.searchBy = params.searchBy || 'title';
    this.filter = params.filter || '';
    this.offset = params.offset || '';
    this.limit = params.limit || '24';
  }

  static toQueryString(params: MovieQueryParams = {}): string {
    const searchParams = new URLSearchParams();
    Object.entries(params).forEach(([key, value]) => {
      if (typeof value !== 'function' && value) {
        searchParams.set(key, value);
      }
    });
    return searchParams.toString();
  }

  static toSearchQueryString(params: MovieQueryParams = {}): string {
    const searchParams = new URLSearchParams();
    if (params.search) {
      searchParams.set('search', params.search);
      return searchParams.toString();
    }
    return '';
  }
}
