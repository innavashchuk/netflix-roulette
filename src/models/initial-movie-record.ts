import { Movie } from './movie';

export interface InitialMovieValues {
  id?: number,
  title: string,
  release_date: Date;
  poster_path: string,
  genres: Array<string>,
  overview: string,
  runtime: number
}

export class InitialMovieRecord implements InitialMovieValues {
  id: number;
  title = '';
  release_date = new Date();
  poster_path = '';
  genres: string[] = [];
  overview = '';
  runtime = 0;

  static mapMovieRecordToInitialValues(record: Movie): InitialMovieValues {
    const initialRecord = new InitialMovieRecord();
    initialRecord.genres = record.genres;
    initialRecord.overview = record.overview;
    if (record.id) {
      initialRecord.id = record.id;
    } else {
      delete initialRecord.id;
    }
    initialRecord.poster_path = record.poster_path;
    initialRecord.release_date = new Date(record.release_date);
    initialRecord.runtime = record.runtime;
    initialRecord.title = record.title;
    return initialRecord;
  }

  static mapInitialValuesToMovieRecord(record: Movie, initialRecord: InitialMovieValues): Movie {
    record.genres = initialRecord.genres;
    record.overview = initialRecord.overview;
    record.poster_path = initialRecord.poster_path;
    record.release_date = initialRecord.release_date.toISOString();
    record.runtime = initialRecord.runtime;
    record.title = initialRecord.title;
    return record;
  }
}
