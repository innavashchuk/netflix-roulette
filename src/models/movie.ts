import { MovieGenreEnum } from './enums/movie-genre';
import { InitialMovieValues } from './initial-movie-record';

export interface Movie {
  id?: number,
  title: string,
  release_date: string;
  poster_path: string,
  genres: Array<string>,
  overview: string,
  runtime: number,
  vote_average: number,
  vote_count: number,
  revenue: number,
  budget: number,
  tagline?: string
}

export class MovieRecord implements Movie {
  title = '';

  release_date = new Date().toISOString();

  poster_path = '';

  genres: string[] = [];

  overview = 'No overview provided';

  runtime = 0;

  vote_average: number;

  vote_count: number;

  revenue: number;

  budget: number;

  tagline: 'No tagline provided';

  constructor(values?: InitialMovieValues) {
    if (values) {
      this.title = values.title;
      this.genres = values.genres;
      this.overview = values.overview;
      this.poster_path = values.poster_path;
      this.release_date = values.release_date.toISOString();
      this.runtime = values.runtime;
    }
  }
}
