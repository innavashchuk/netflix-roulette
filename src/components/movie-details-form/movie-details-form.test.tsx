import * as React from 'react';
import {render, screen, fireEvent, within, waitFor} from '@testing-library/react';
import MovieDetailsForm from './Movie-Details-Form';
import {InitialMovieRecord, InitialMovieValues} from '../../models/initial-movie-record';

const DATE_STRING = '01/01/2001';
const MOVIE: InitialMovieValues = {
    'title': 'Star Wars: The Last Jedi',
    'release_date': new Date(DATE_STRING),
    'poster_path': 'https://image.tmdb.org/t/p/w500/kOVEVeg59E0wsnXmF9nrh6OmWII.jpg',
    'overview': 'Rey develops her newly discovered abilities with the guidance of Luke Skywalker, who is unsettled by the strength of her powers. Meanwhile, the Resistance prepares to do battle with the First Order.',
    'genres': [
        'Fantasy'
    ],
    'runtime': 152
};

const EMPTY_MOVIE = new InitialMovieRecord();

const onFormSubmit = jest.fn();

describe('MovieDetailsForm', () => {
    it('should initialize', () => {
        const {asFragment} = render(<MovieDetailsForm movieRecord={MOVIE} onFormSubmit={onFormSubmit} isEdit={false}/>);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should display EDIT in title if isEdit is true', () => {
        render(<MovieDetailsForm movieRecord={MOVIE} onFormSubmit={onFormSubmit} isEdit/>);
        const title = screen.getByRole('heading');
        expect(title.textContent).toEqual('EDIT MOVIE');
    });

    it('should display ADD in title if isEdit is false', () => {
        render(<MovieDetailsForm movieRecord={MOVIE} onFormSubmit={onFormSubmit} isEdit={false}/>);
        const title = screen.getByRole('heading');
        expect(title.textContent).toEqual('ADD MOVIE');
    });

    it('should reset form when RESET button is clicked', () => {
        const {getByRole} = render(<MovieDetailsForm movieRecord={MOVIE} onFormSubmit={onFormSubmit} isEdit/>);
        const resetButton = getByRole('button', {name: 'RESET'});
        const nameTextField = getByRole('textbox', {name: 'TITLE'});
        const newTitle = 'New Title';
        fireEvent.change(nameTextField, {
            target: {
                value: newTitle
            }
        });
        expect(nameTextField.getAttribute('value')).toEqual(newTitle);

        fireEvent.click(resetButton);
        expect(nameTextField.getAttribute('value')).toEqual(MOVIE.title);
    });

    it('should validate TITLE', async () => {
        const {getByRole, getByText} = render(<MovieDetailsForm movieRecord={EMPTY_MOVIE} onFormSubmit={onFormSubmit}
                                                                isEdit/>);
        const nameTextField = getByRole('textbox', {name: 'TITLE'});
        expect(nameTextField.getAttribute('value')).toBeFalsy();
        const submitButton = getByRole('button', {name: 'SUBMIT'});
        fireEvent.click(submitButton);
        await waitFor(() => {
            const errorMessage = getByText((content, element) => content.includes('Title is required'));
            expect(errorMessage).toBeTruthy();
        });
    });

});

test('should call onFormSubmit with changed record when submit button is clicked', async () => {
    const {getByRole, getByPlaceholderText} = render(<MovieDetailsForm movieRecord={EMPTY_MOVIE}
                                                                       onFormSubmit={onFormSubmit} isEdit={false}/>);
    const submitButton = getByRole('button', {name: 'SUBMIT'});
    const nameTextField = getByRole('textbox', {name: 'TITLE'});
    fireEvent.change(nameTextField, {
        target: {
            value: MOVIE.title
        }
    });

    const releaseDateField = getByPlaceholderText('Select Date');
    fireEvent.change(releaseDateField, {
        target: {
            value: DATE_STRING
        }
    });

    const runtimeField = getByRole('spinbutton', {
        name: 'RUNTIME'
    });
    fireEvent.change(runtimeField, {
        target: {
            value: MOVIE.runtime
        }
    });

    const urlField = getByRole('textbox', {
        name: 'MOVIE URL'
    });
    fireEvent.change(urlField, {
        target: {
            value: MOVIE.poster_path
        }
    });

    const overviewField = getByRole('textbox', {
        name: 'OVERVIEW'
    });
    fireEvent.change(overviewField, {
        target: {
            value: MOVIE.overview
        }
    });

    const genresField = getByRole('button', {
        name: 'GENRES'
    });
    fireEvent.mouseDown(genresField);
    const listbox = within(getByRole('listbox'));
    expect(listbox).toBeTruthy();
    fireEvent.mouseDown(getByRole('option', {
        name: 'Fantasy'
    }));
    fireEvent.click(getByRole('option', {
        name: 'Fantasy'
    }));

    expect(submitButton).toBeTruthy();
    expect(submitButton.getAttribute('disabled')).toBeFalsy();
    fireEvent.click(submitButton);
    await waitFor(() =>
        expect(onFormSubmit).toHaveBeenCalledWith(MOVIE)
    )
});
