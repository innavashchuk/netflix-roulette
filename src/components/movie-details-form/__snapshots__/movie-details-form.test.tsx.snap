// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetailsForm should initialize 1`] = `
<DocumentFragment>
  <div
    class="movie-details"
  >
    <h1>
      ADD MOVIE
    </h1>
    <form
      action="#"
      autocomplete="off"
      class="movie-details__form"
    >
      <label
        class="form-label"
        for="title"
      >
        TITLE
      </label>
      <input
        aria-label="TITLE"
        class="form-input"
        name="title"
        placeholder="Title here"
        type="text"
        value="Star Wars: The Last Jedi"
      />
      <label
        class="form-label"
        for="release_date"
      >
        RELEASE DATE
      </label>
      <div
        aria-label="RELEASE DATE"
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
            id="date-picker-inline"
            name="release_date"
            placeholder="Select Date"
            type="text"
            value="01/01/2001"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            <button
              aria-label="change date"
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                  />
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <label
        class="form-label"
        for="poster_path"
      >
        MOVIE URL
      </label>
      <input
        aria-label="MOVIE URL"
        class="form-input"
        name="poster_path"
        placeholder="Movie URL here"
        type="text"
        value="https://image.tmdb.org/t/p/w500/kOVEVeg59E0wsnXmF9nrh6OmWII.jpg"
      />
      <label
        class="form-label"
        for="genres"
      >
        GENRE
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline form-input "
      >
        <div
          aria-haspopup="listbox"
          aria-label="GENRES"
          aria-labelledby="genres"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="genres"
          role="button"
          tabindex="0"
        >
          Fantasy
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          name="genres"
          tabindex="-1"
          value="Fantasy"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <label
        class="form-label"
        for="overview"
      >
        OVERVIEW
      </label>
      <textarea
        aria-label="OVERVIEW"
        class="form-input "
        name="overview"
        placeholder="Overview here"
        rows="1"
        style="height: -4px; overflow: hidden;"
      >
        Rey develops her newly discovered abilities with the guidance of Luke Skywalker, who is unsettled by the strength of her powers. Meanwhile, the Resistance prepares to do battle with the First Order.
      </textarea>
      <textarea
        aria-hidden="true"
        class="form-input "
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0);"
        tabindex="-1"
      />
      <label
        class="form-label"
        for="runtime"
      >
        RUNTIME
      </label>
      <input
        aria-label="RUNTIME"
        class="form-input"
        name="runtime"
        placeholder="Runtime here"
        type="number"
        value="152"
      />
      <div
        class="movie-details__btns"
      >
        <button
          class="button-secondary"
          name="reset"
          type="reset"
        >
          RESET
        </button>
        <button
          class="button-primary"
          name="submit"
          type="submit"
        >
          SUBMIT
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
